### PR TITLE
Add metrics module with binary f1, precision, recall and dice

### PR DESCRIPTION
**Pull request by blazekadam**
PR includes tests.

### CODE REVIEW

File: cxflow_tf/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
 from .tf_net import BaseTFNet, BaseTFNetRestore
 from .utils import TF_OPTIMIZERS_MODULE, create_activation, create_optimizer, repeat
 from .hooks import *
+from .metrics import *
```
Comment (FloopCZ) [inline cxflow_tf/__init__.py]: Me and @petrbel, we would prefer explicit naming to avoid namespace pollution.

File: cxflow_tf/metrics.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    The coefficient is defined as 2*|X <intersection> Y| / (|X| + |Y|).
+
+    :param predictions: 2-d tensor (batch, predictions) of predicted 0/1 classes
+    :param labels: 2-d tensor (batch, labels) of expected 0/1 classes
+    :return: batched Sørensen–Dice coefficients
+    """
+    predictions = tf.cast(predictions, tf.int32)
+    labels = tf.cast(labels, tf.int32)
+
+    true_positives = tf.reduce_sum(tf.cast(tf.cast(predictions, tf.bool) & tf.cast(labels, tf.bool), tf.int32), axis=1)
```
Comment (FloopCZ) [inline cxflow_tf/metrics.py]: Again, the casts may be avoided by elementwise multiplication.

File: cxflow_tf/tests/metrics_test.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            self.assertTrue(np.isnan(computed_dice[-1]))
+
+    def test_stats(self):
+        """ Test if f1, precision and recall stats are computed as expected."""
+
+        with tf.Session().as_default():
+            labels = tf.constant(_LABELS, dtype=tf.int32)
+            predictions = tf.constant(_PREDICTIONS, dtype=tf.int32)
+            expected_recall = np.array([1./3, 0.])
+            expected_precision = np.array([3./6, 0.])
```
Comment (FloopCZ) [inline cxflow_tf/tests/metrics_test.py]: Please write as `1./2` :-)

File: cxflow_tensorflow/metrics.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    Calculate f1, precision and recall from binary classification expected and predicted values.
+
+    :param predictions: 2-d tensor (batch, predictions) of predicted 0/1 classes
+    :param labels: 2-d tensor (batch, labels) of expected 0/1 classes
+    :return: a tuple of batched (f1, precision and recall) values
+    """
+    predictions = tf.cast(predictions, tf.int32)
+    labels = tf.cast(labels, tf.int32)
+
+    true_positives = tf.reduce_sum(tf.cast(tf.cast(predictions, tf.bool) & tf.cast(labels, tf.bool), tf.int32), axis=1)
```
Comment (FloopCZ) [inline cxflow_tensorflow/metrics.py]: Maybe we could avoid two casts by a simple multiplication of the original arrays (they are of type tf.int32).

File: cxflow_tensorflow/metrics.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    The coefficient is defined as 2*|X <intersection> Y| / (|X| + |Y|).
+
+    :param predictions: 2-d tensor (batch, predictions) of predicted 0/1 classes
+    :param labels: 2-d tensor (batch, labels) of expected 0/1 classes
+    :return: batched Sørensen–Dice coefficients
+    """
+    predictions = tf.cast(predictions, tf.int32)
+    labels = tf.cast(labels, tf.int32)
+
+    true_positives = tf.reduce_sum(tf.cast(tf.cast(predictions, tf.bool) & tf.cast(labels, tf.bool), tf.int32), axis=1)
```
Comment (FloopCZ) [inline cxflow_tensorflow/metrics.py]: Again, the casts may be avoided by elementwise multiplication.

File: cxflow_tensorflow/__init__.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
 from .net import BaseNet
 from .utils import TF_OPTIMIZERS_MODULE, create_activation, create_optimizer, repeat
 from .hooks import *
+from .metrics import *
```
Comment (FloopCZ) [inline cxflow_tensorflow/__init__.py]: Me and @petrbel, we would prefer explicit naming to avoid namespace pollution.

From @petrbel:
The list of imported entities should be listed explicitely. This may import also unwanted constatns or imported entities from the nested file.

File: cxflow_tensorflow/tests/metrics_test.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            self.assertTrue(np.isnan(computed_dice[-1]))
+
+    def test_stats(self):
+        """ Test if f1, precision and recall stats are computed as expected."""
+
+        with tf.Session().as_default():
+            labels = tf.constant(_LABELS, dtype=tf.int32)
+            predictions = tf.constant(_PREDICTIONS, dtype=tf.int32)
+            expected_recall = np.array([1./3, 0.])
+            expected_precision = np.array([3./6, 0.])
```
Comment (FloopCZ) [inline cxflow_tensorflow/tests/metrics_test.py]: Please write as 1./2 (one of two 1's was right) :-)

File: cxflow_tensorflow/metrics.py [hunk 8]
```diff
@@ -15,7 +15,7 @@ def bin_stats(predictions, labels):
     predictions = tf.cast(predictions, tf.int32)
     labels = tf.cast(labels, tf.int32)
 
-    true_positives = tf.reduce_sum(tf.cast(tf.cast(predictions, tf.bool) & tf.cast(labels, tf.bool), tf.int32), axis=1)
+    true_positives = tf.reduce_sum((predictions * labels), axis=1)
```
Comment (FloopCZ) [inline cxflow_tensorflow/metrics.py]: Are the braces around `(predictions * labels)` necessary?

Comment (blazekadam) [inline cxflow_tensorflow/metrics.py]: apparently, they are

**Comment by FloopCZ** [comment]
Thank you @blazekadam, it will come in handy.
